### PR TITLE
[Refactor] Replan the insert to get rid of the metadata inconsistent between plan() and beginTransaction()

### PR DESCRIPTION
**Pull request by chaoyli**
The metadata may be changed between plan() and beginTransaction(). When to beginTransaction(), the plan is not ready and the tablet is dropped by balanced. So we need to get txn first, to make the dropping tablet procedure to wait.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java [hunk 1]
```diff
@@ -487,7 +487,7 @@ public void setTopProfileSupplier(Supplier<RuntimeProfile> topProfileSupplier) {
         this.topProfileSupplier = topProfileSupplier;
     }
 
-    public void setExecPlanSupplier(Supplier<ExecPlan> execPlanSupplier) {
+    public void setExecPlanSupplier(ExecPlan execPlanSupplier) {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:490]:
```suggestion
    public void setExecPlanSupplier(ExecPlan execPlan) {
```

File: fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java [hunk 2]
```diff
@@ -1700,7 +1701,7 @@ public void handleDMLStmt(ExecPlan execPlan, DmlStmt stmt) throws Exception {
             QeProcessorImpl.INSTANCE.registerQuery(context.getExecutionId(), queryInfo);
             coord.exec();
             coord.setTopProfileSupplier(this::buildTopLevelProfile);
-            coord.setExecPlanSupplier(() -> execPlan);
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java:1703]: why change the functional interface? you can keep it, when get() is called, i think i will return the new one

File: fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java [hunk 3]
```diff
@@ -1619,6 +1619,7 @@ public void handleDMLStmt(ExecPlan execPlan, DmlStmt stmt) throws Exception {
                             FrontendOptions.getLocalHostAddress()),
                     sourceType,
                     context.getSessionVariable().getQueryTimeoutS());
+            execPlan = StatementPlanner.plan(stmt, context);
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java:1622]: how about add comments explaining why you re-plan?

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 5 / 8 (62.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|2|5|40.00%|[1542, 1916, 1917]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java [hunk 4]
```diff
@@ -1620,6 +1620,13 @@ public void handleDMLStmt(ExecPlan execPlan, DmlStmt stmt) throws Exception {
                     sourceType,
                     context.getSessionVariable().getQueryTimeoutS());
 
+            // The metadata may be changed between plan() and beginTransaction().
+            // When to beginTransaction(), the plan is not ready and the tablet is dropped by balance.
+            // So we need to get txn first, to make the dropping tablet procedure to wait.
+            // This is the re-plan of the insert to fix the issue
+
+            execPlan = StatementPlanner.plan(stmt, context);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java:1628]: could we only plan once?

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by nshangyiming — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java (hunk 4 above)
Comment (chaoyli) [reply fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java]: It's a good idea. Currently, we should place the begin_transaction() in front of any request. It needs much refactoring of the transaction framework. I have added a configuration for this.

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 6 / 10 (60.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|2|5|40.00%|[1542, 1916, 1917]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|3|4|75.00%|[1629]|
|:large_blue_circle:|com/starrocks/common/Config.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**